server/src/main/java/org/apache/iotdb/db/engine/tsfilemanagement/utils/HotCompactionUtils.java

```diff
@@ -91,130 +84,204 @@ private HotCompactionUtils() {
-          hotCompactionLogger.logDevice(deviceId, writer.getPos());
+          hotCompactionLogger.logDevice(device, writer.getPos());
         }
-      }
-    } else {
-      for (Entry<String, Map<String, MeasurementSchema>> deviceMeasurementEntry : deviceMeasurementMap
-          .entrySet()) {
-        String deviceId = deviceMeasurementEntry.getKey();
-        writer.startChunkGroup(deviceId);
-        for (Entry<String, MeasurementSchema> entry : deviceMeasurementEntry.getValue()
+      } else {
+        long maxVersion = Long.MIN_VALUE;
+        for (Entry<String, Map<TsFileSequenceReader, List<ChunkMetadata>>> entry : measurementChunkMetadataMap
             .entrySet()) {
-          String measurementId = entry.getKey();
-          Pair<ChunkMetadata, Chunk> chunkPair = writeSeqChunk(storageGroup,
-              tsFileSequenceReaderMap, deviceId, measurementId, tsFileResources);
-          ChunkMetadata newChunkMetadata = chunkPair.left;
-          Chunk newChunk = chunkPair.right;
-          if (newChunkMetadata != null && newChunk != null) {
-            // wait for limit write
-            MergeManager.mergeRateLimiterAcquire(compactionRateLimiter,
-                (long)newChunk.getHeader().getDataSize() + newChunk.getData().position());
-            writer.writeChunk(newChunk, newChunkMetadata);
-            targetResource.updateStartTime(deviceId, newChunkMetadata.getStartTime());
-            targetResource.updateEndTime(deviceId, newChunkMetadata.getEndTime());
+          Map<TsFileSequenceReader, List<ChunkMetadata>> readerChunkMetadatasMap = entry.getValue();
+          boolean isPageEnoughLarge = true;
+          for (List<ChunkMetadata> chunkMetadatas : readerChunkMetadatasMap.values()) {
+            for (ChunkMetadata chunkMetadata : chunkMetadatas) {
+              if (chunkMetadata.getNumOfPoints() < mergePagePointNum) {
+                isPageEnoughLarge = false;
+                break;
+              }
+            }
+          }
+          if (isPageEnoughLarge) {
+            logger.info("{} [Hot Compaction] page enough large, use append merge", storageGroup);
+            maxVersion = writeByAppendMerge(maxVersion, device, compactionWriteRateLimiter,
+                compactionReadRateLimiter,
+                readerChunkMetadatasMap, targetResource, writer);
+          } else {
+            logger
+                .info("{} [Hot Compaction] page enough large, use deserialize merge", storageGroup);
+            maxVersion = writeByDeserializeMerge(maxVersion, device, compactionWriteRateLimiter,
+                compactionReadRateLimiter,
+                entry,
+                targetResource, writer);
           }
         }
         writer.endChunkGroup();
+        writer.writeVersion(maxVersion);
         if (hotCompactionLogger != null) {
-          hotCompactionLogger.logDevice(deviceId, writer.getPos());
+          hotCompactionLogger.logDevice(device, writer.getPos());
```

Comment:
This code is consistent with the if, can be extracted

Reply:
good idea